graphs/bi_directional_dijkstra.py

```diff
@@ -0,0 +1,142 @@
+
+    shortest_distance = np.inf
+
+    queue_forward.put((0, source))
+    queue_backward.put((0, destination))
+
+    if source == destination:
+        return 0
+
+    while queue_forward and queue_backward:
+        while not queue_forward.empty():
+            _, v_fwd = queue_forward.get()
+
+            if v_fwd not in visited_forward:
+                break
+        else:
+            break
+        visited_forward.add(v_fwd)
+
+        while not queue_backward.empty():
+            _, v_bwd = queue_backward.get()
+
+            if v_bwd not in visited_backward:
+                break
+        else:
+            break
+        visited_backward.add(v_bwd)
+
+        # forward pass and relaxation
+        for nxt_fwd, d_forward in graph_forward[v_fwd]:
+            if nxt_fwd in visited_forward:
+                continue
+            old_cost_f = cst_fwd.get(nxt_fwd, np.inf)
+            new_cost_f = cst_fwd[v_fwd] + d_forward
+            if new_cost_f < old_cost_f:
+                queue_forward.put((new_cost_f, nxt_fwd))
+                cst_fwd[nxt_fwd] = new_cost_f
+                parent_forward[nxt_fwd] = v_fwd
+            if nxt_fwd in visited_backward:
+                if cst_fwd[v_fwd] + d_forward + cst_bwd[nxt_fwd] < shortest_distance:
+                    shortest_distance = cst_fwd[v_fwd] + d_forward + cst_bwd[nxt_fwd]
+
+        # backward pass and relaxation
+        for nxt_bwd, d_backward in graph_backward[v_bwd]:
+            if nxt_bwd in visited_backward:
+                continue
+            old_cost_b = cst_bwd.get(nxt_bwd, np.inf)
+            new_cost_b = cst_bwd[v_bwd] + d_backward
+            if new_cost_b < old_cost_b:
+                queue_backward.put((new_cost_b, nxt_bwd))
+                cst_bwd[nxt_bwd] = new_cost_b
+                parent_backward[nxt_bwd] = v_bwd
+
+            if nxt_bwd in visited_forward:
+                if cst_bwd[v_bwd] + d_backward + cst_fwd[nxt_bwd] < shortest_distance:
+                    shortest_distance = cst_bwd[v_bwd] + d_backward + cst_fwd[nxt_bwd]
+
+        if cst_fwd[v_fwd] + cst_bwd[v_bwd] >= shortest_distance:
+            break
+
+    if shortest_distance == np.inf:
+        return shortest_path_distance
+    shortest_path_distance = shortest_distance
+    return shortest_path_distance
+
+
+r"""
+Layout of input Graph:
+E -- 1 --> B -- 1 --> C -- 1 --> D -- 1 --> F
+ \                                         /\
+  \                                        ||
+    -------- 2 ---------> G ------- 1 ------
+"""
+
+graph_fwd = {
+    "B": [["C", 1]],
+    "C": [["D", 1]],
+    "D": [["F", 1]],
+    "E": [["B", 1], ["G", 2]],
+    "F": [],
+    "G": [["F", 1]],
+}
+graph_bwd = {
+    "B": [["E", 1]],
+    "C": [["B", 1]],
+    "D": [["C", 1]],
+    "F": [["D", 1], ["G", 1]],
+    "G": [["E", 2]],
+}
+print(
+    bidirectional_dij("E", "F", graph_fwd, graph_bwd)
+)  # # E -- 2 --> G -- 1 --> F == 3
+
+if __name__ == "__main__":
```

Comment:
This should be moved within the if name == main

Reply:
Cannot put `graph_fwd ` and `graph_bwd` inside any other scope because it was giving error in the GitHub workflow while running `test` that  `graph_fwd ` is not defined

Reply:
<img width="865" alt="Screenshot 2022-11-14 at 5 28 41 PM" src="https://user-images.githubusercontent.com/74960567/201654786-d575be45-fa16-45ed-bfd6-2bf8f91b5a6c.png">